File: JAVA/GestionRelationEmployees/src/main/java/models/ProjectUpdate.java
package models;

import java.sql.Timestamp;

/**
 * Modèle représentant une mise à jour/activité sur un projet
 */
public class ProjectUpdate {

    // Enums
    public enum UpdateType {
        status_change, milestone, task, comment, document
    }

    private int id;
    private int projectId;
    private int userId;
    private UpdateType updateType;
    private String title;
    private String content;
    private Timestamp createdAt;

    // Champs transients pour l'affichage
    private String projectName;
    private String username;

    // ─── Constructeurs ───────────────────────────────────────────────

    public ProjectUpdate() {}

    public ProjectUpdate(int projectId, int userId, UpdateType updateType,
                         String title, String content) {
        this.projectId = projectId;
        this.userId = userId;
        this.updateType = updateType;
        this.title = title;
        this.content = content;
    }

    // ─── Getters & Setters ───────────────────────────────────────────

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public int getProjectId() { return projectId; }
    public void setProjectId(int projectId) { this.projectId = projectId; }

    public int getUserId() { return userId; }
    public void setUserId(int userId) { this.userId = userId; }

    public UpdateType getUpdateType() { return updateType; }
    public void setUpdateType(UpdateType updateType) { this.updateType = updateType; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }

    public Timestamp getCreatedAt() { return createdAt; }
    public void setCreatedAt(Timestamp createdAt) { this.createdAt = createdAt; }

    // Champs transients
    public String getProjectName() { return projectName; }
    public void setProjectName(String projectName) { this.projectName = projectName; }

    public String getUsername() { return username; }
    public void setUsername(String username) { this.username = username; }

    // ─── Méthodes utilitaires ────────────────────────────────────────

    /**
     * Retourne un emoji de type pour l'affichage
     */
    public String getTypeEmoji() {
        return switch (updateType) {
            case status_change -> "🔄";
            case milestone -> "🎯";
            case task -> "✓";
            case comment -> "💬";
            case document -> "📎";
        };
    }

    /**
     * Retourne le nom du type pour l'affichage
     */
    public String getTypeLabel() {
        return switch (updateType) {
            case status_change -> "Changement de statut";
            case milestone -> "Jalon";
            case task -> "Tâche";
            case comment -> "Commentaire";
            case document -> "Document";
        };
    }

    /**
     * Retourne une version formatée de la date
     */
    public String getFormattedDate() {
        if (createdAt == null) return "";

        long diff = System.currentTimeMillis() - createdAt.getTime();
        long minutes = diff / (1000 * 60);
        long hours = diff / (1000 * 60 * 60);
        long days = diff / (1000 * 60 * 60 * 24);

        if (minutes < 1) return "À l'instant";
        if (minutes < 60) return "Il y a " + minutes + " min";
        if (hours < 24) return "Il y a " + hours + "h";
        if (days == 1) return "Hier";
        if (days < 7) return "Il y a " + days + " jours";

        return createdAt.toString().substring(0, 16);  // YYYY-MM-DD HH:MM
    }

    /**
     * Retourne un résumé court pour la liste d'activités
     */
    public String getSummary() {
        String emoji = getTypeEmoji();
        String date = getFormattedDate();
        String user = username != null ? username : "Utilisateur";

        return emoji + " " + date + " - " + user + ": " + title;
    }

    @Override
    public String toString() {
        return "ProjectUpdate{id=" + id + ", type=" + updateType + ", title='" + title + "'}";
    }
}

File: src/Entity/Projet/ProjectUpdate.php
<?php

namespace App\Entity\Projet;

use App\Repository\Projet\ProjectUpdateRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: ProjectUpdateRepository::class)]
#[ORM\Table(name: 'project_updates')]
#[ORM\Index(name: 'idx_project_updates_project_id', columns: ['project_id'])]
#[ORM\Index(name: 'idx_project_updates_created_at', columns: ['created_at'])]
#[ORM\HasLifecycleCallbacks]
class ProjectUpdate
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(name: 'project_id', type: Types::INTEGER)]
    private ?int $projectId = null;

    #[ORM\Column(name: 'user_id', type: Types::INTEGER)]
    private ?int $userId = null;

    #[ORM\Column(name: 'update_type', length: 40)]
    private ?string $updateType = null;

    #[ORM\Column(length: 255)]
    private ?string $title = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    private ?string $content = null;

    #[ORM\Column(name: 'created_at', type: Types::DATETIME_MUTABLE)]
    private \DateTimeInterface $createdAt;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function __construct()
    {
        $this->createdAt = new \DateTime();
    }

    public function getProjectId(): ?int
    {
        return $this->projectId;
    }

    public function setProjectId(int $projectId): static
    {
        $this->projectId = $projectId;
        return $this;
    }

    public function getUserId(): ?int
    {
        return $this->userId;
    }

    public function setUserId(int $userId): static
    {
        $this->userId = $userId;
        return $this;
    }

    public function getUpdateType(): ?string
    {
        return $this->updateType;
    }

    public function setUpdateType(string $updateType): static
    {
        $this->updateType = $updateType;
        return $this;
    }

    public function getTitle(): ?string
    {
        return $this->title;
    }

    public function setTitle(string $title): static
    {
        $this->title = $title;
        return $this;
    }

    public function getContent(): ?string
    {
        return $this->content;
    }

    public function setContent(?string $content): static
    {
        $this->content = $content;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    #[ORM\PrePersist]
    public function onPrePersist(): void
    {
        $this->createdAt = new \DateTime();
    }
}

